pkg/tests/install/test.py

```diff
@@ -0,0 +1,142 @@
+            # We can't infer what TreeArtifacts are by looking at them -- the
+            # build system is not aware of their contents.
+            raise ValueError("Entity {} is not a link, file, or directory")
+
+    def assertEntryTypeMatches(self, entry, actual_path):
+        actual_entry_type = self.entity_type_at_path(actual_path)
+        self.assertEqual(actual_entry_type, entry.entry_type,
+                        "Entity {} should be a {}, but was actually {}".format(
+                            entry.dest,
+                            entry_type_to_string(entry.entry_type),
+                            entry_type_to_string(actual_entry_type),
+                        ))
+
+    def assertEntryModeMatches(self, entry, actual_path):
+        # TODO: permissions in windows are... tricky.  Don't bother
+        # testing for them if we're in it for the time being
+        if os.name == 'nt':
+            return
+
+        actual_mode = stat.S_IMODE(os.stat(actual_path).st_mode)
+        expected_mode = int(entry.mode, 8)
+        self.assertEqual(actual_mode, expected_mode,
+                         "Entry {} has mode {:04o}, expected {:04o}".format(
+                            entry.dest, actual_mode, expected_mode,
+                        ))
+
+    def test_manifest_matches(self):
+        # TODO: check for ownership (user, group)
+        found_entries = {dest: False for dest in self.manifest_data.keys()}
+        for root, dirs, files in os.walk(self.installdir):
+            rel_root_path = os.path.relpath(root, self.installdir)
+
+            # TODO(nacl): check for treeartifacts here.  If so, prune `dirs`,
+            # and set the rest aside for future processing.
+
+            # Directory ownership tests
+            if len(files) == 0 and len(dirs) == 0:
+                # Empty directories must be explicitly requested by something
+                if rel_root_path not in self.manifest_data:
+                    self.fail("Directory {} not in manifest".format(rel_root_path))
+
+                entry = self.manifest_data[rel_root_path]
+                self.assertEntryTypeMatches(entry, root)
+                self.assertEntryModeMatches(entry, root)
+
+                found_entries[rel_root_path] = True
+            else:
+                # There's something in here.  Depending on how it was set up, it
+                # could either be owned or unowned.
+                if rel_root_path in self.manifest_data:
+                    entry = self.manifest_data[rel_root_path]
+                    self.assertEntryTypeMatches(entry, root)
+                    self.assertEntryModeMatches(entry, root)
+
+                    found_entries[rel_root_path] = True
+                else:
+                    print("{} is unowned".format(rel_root_path))
+
+            for f in files:
+                # The path on the filesystem in which the file actually exists.
+                fpath = os.path.normpath("/".join([root, f]))
```

Comment:
Should that be os.path.sep instead of "/"

Reply:
This ends up being a key into the `manifest_data` dict, which, to the best of my knowledge, will always have UNIX-style path separators.

Added a comment mentioning #387.